extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -860,8 +835,456 @@ private Table readTableInternal(
+                                    "provided when writing multiple tables with different definitions");
+                }
+            }
+            useDefinition = firstDefinition;
+            writeInstructions = writeInstructions.withTableDefinition(useDefinition);
+        }
+
+        // Try loading the table from the catalog, or create if required
+        final org.apache.iceberg.Table icebergTable;
+        final SpecAndSchema newSpecAndSchema;
+        final boolean newNamespaceCreated;
+        final boolean newTableCreated;
+        if (catalog.tableExists(tableIdentifier)) {
+            icebergTable = catalog.loadTable(tableIdentifier);
+            newSpecAndSchema = createSpecAndSchema(useDefinition, writeInstructions);
+            newNamespaceCreated = false;
+            newTableCreated = false;
+            if (verifySchema) {
+                try {
+                    if (overwrite) {
+                        verifyOverwriteCompatibility(icebergTable.schema(), newSpecAndSchema.schema());
+                        verifyOverwriteCompatibility(icebergTable.spec(), newSpecAndSchema.partitionSpec());
+                    } else {
+                        verifyAppendCompatibility(icebergTable.schema(), useDefinition, writeInstructions);
+                        verifyAppendCompatibility(icebergTable.spec(), useDefinition, writeInstructions);
+                    }
+                } catch (final IllegalArgumentException e) {
+                    throw new IllegalArgumentException("Schema verification failed. Please provide a compatible " +
+                            "schema or disable verification in the Iceberg instructions. See the linked exception " +
+                            "for more details.", e);
+                }
+            }
+        } else if (writeInstructions.createTableIfNotExist()) {
+            newNamespaceCreated = createNamespaceIfNotExists(tableIdentifier.namespace());
+            newSpecAndSchema = createSpecAndSchema(useDefinition, writeInstructions);
+            icebergTable = createNewIcebergTable(tableIdentifier, newSpecAndSchema, writeInstructions);
+            newTableCreated = true;
+        } else {
+            throw new IllegalArgumentException(
+                    "Table does not exist: " + tableIdentifier + ", update the instructions " +
+                            "to create the table if it does not exist and try again.");
+        }
+
+        try {
+            final List<CompletedParquetWrite> parquetFileinfo =
+                    writeParquet(icebergTable, dhTables, writeInstructions);
+            final List<DataFile> appendFiles = dataFilesFromParquet(parquetFileinfo);
+            if (addSnapshot) {
+                commit(icebergTable, newSpecAndSchema, appendFiles, overwrite && !newTableCreated, verifySchema);
+            }
+            return appendFiles;
+        } catch (final RuntimeException writeException) {
```

Comment:
Maybe we should make this broader? `Throwable`? Or just stay out of the way of any `Error`?